js/classes/entities/midground/enemies/ground_enemy_base.js

```diff
@@ -0,0 +1,199 @@
+ * Handles movement, block collisions, etc.
+ * 
+ * @author Trae Claar
+ */
+class GroundEnemyBase {
+
+    /**
+     * Constructor for a GroundEnemyBase.
+     * 
+     * @param {Entity} enemy the enemy associated with this EnemyBase
+     * @param {Vector} pos the initial position of the enemy
+     * @param {number} scaledSize the size of the enemy
+     * @param {number} speed the speed of the enemy
+     * @param {number} health the initial health of the enemy
+     * @param {number} maxRoamDistance the maximum distance the enemy will wander 
+     *      from its initial position when roaming
+     * @param {function} onDeath callback function invoked when the enemy dies
+     * @param {string} stance the enemy's stance (use an EnemyBase member stance, 
+     *      e.g. EnemyBase.AggressiveStance)
+     * @param {number} reactionDistance the maximum distance that the enemy will 
+     *      react to Chad according to its stance and state
+     */
+    constructor(enemy, pos, scaledSize, speed, health, maxRoamDistance, onDeath, stance, reactionDistance) {
+        this.enemy = enemy;
+        this.base = new EnemyBase(enemy, pos, scaledSize, speed, health, onDeath, () => this.handleDamage());
+        
+        enemy.state = "roam";
+        enemy.action = "idle";
+        enemy.maxRoamDistance = maxRoamDistance; // *
+        enemy.yVelocity = 0;
+
+        enemy.getDirection = () => this.getDirection();
+
+        this.minRoamX = pos.x - maxRoamDistance;
+        this.targetX = pos.x;
+        this.stance = stance;
+        this.reactionDistance = reactionDistance ?? Camera.SIZE.x / 2 - scaledSize.x;
+
+        // for backwards compatibility with enemy classes that expect the base to have these methods
+        this.getFacing = enemy.getFacing; 
+        this.getCenter = enemy.getCenter;
+    }
+
+    /** 
+     * Aggressive enemy stance. Aggressive enemies will pursue Chad any time he
+     * is within their maximum reaction distance. 
+     */
+    static get AGGRESSIVE_STANCE() {
+        return "aggressive";
+    }
+
+    /** 
+     * Defensive enemy stance. Defensive enemies will pursue/attack Chad only when
+     * provoked.
+     */
+    static get DEFENSIVE_STANCE() {
+        return "defensive";
+    }
+
+    /** Passive enemy stance. Passive enemies will run from Chad when attacked. */
+    static get PASSIVE_STANCE() {
+        return "passive";
+    }
+
+    /** 
+     * Avoid enemy stance. Enemies with the avoid stance will run from Chad whenever
+     * he is within their maximum reaction distance.
+     */
+    static get AVOID_STANCE() {
+        return "avoid";
+    }
+
+    /** The initial velocity, in pixels/s of an enemy's jump. */
+    get jumpVelocity() {
+        return 450;
+    }
+    
+    /**
+     * Make the enemy react to taking damage.
+     */
+    handleDamage() {
+        switch (this.stance) {
+            case GroundEnemyBase.AGGRESSIVE_STANCE:
```

Comment:
I didn't know you could have multiple cases without a break. Curious how that works here. Will an enemy be aggressive for a while then defensive? Why are there not breaks between each case?

Reply:
This allows for the same behavior in multiple cases, e.g. aggressive and defensive both set state to pursue in this method.